metaflow/plugins/kfp/kfp.py

```diff
@@ -564,6 +565,49 @@ def _step_cli(self, node: DAGNode, task_id: str, user_code_retries: int) -> str:
         cmds.append(" ".join(entrypoint + top_level + step))
         return " && ".join(cmds)
 
+    @staticmethod
+    def _create_resource_based_node_type_toleration(
+        cpu: float, memory: float
+    ) -> Optional[V1Toleration]:
+        """Allow large enough pod to use higher cost nodes
+        TODO: Replace using AIP-5264 MutatingWebHook (or Validating) for AIP pod scheduling policies
+
+        The following node types are considered for setting the threshold:
+        c5.4xlarge: 16 vCPU, 32 GB
+        r5.12xlarge: 48 vCPU, 384 GB
+        Resource threshold are lower than machine resource boundary to take overheads into
+        account.
+
+        Toleration allows pods to utilize larger nodes without enforcement.
+        Setting a low threshold for using larger host allow more freedom for scheduler
+        and potentially higher utilization rate.
+
+        cpu: number of vCPU requested. Fractions are allowed.
+        memory: memory requested in GB (not GiB)
+        """
+        # Base on observed resource data Oct. 21, 2021, available resource per c5.4xlarge node:
+        # Memory: 20.24GB = 18.85 GiB = 27.72 GiB (allocatable) - 8.87 GiB (DaemonSet)
+        # CPU: 10.34 vCPU = 15.89 (allocatable) - 5.55 (DaemonSet)
+        # Argo additionally adds a "wait" container to pods per step, taking default resources
+
+        # Threshold should leave significant margin below estimated available resource
+        # Pods will be unschedulable if its resource requirement falls in range
+        #   [available resource, thresholdfor large node)
+
+        # Using 50% of total available resource on c5.4xlarge
+        memory_threshold = 16  # GB
```

Comment:
Unless I'm missing something, shouldn't this `memory_threshold` be `10` = `20.24GB / 2 = ~10GB`.

Reply:
Oh I should make the doc string more clear - by total available I mean 50% of the machine resource. 
16G is already 4 G under the available resource.

I was thinking about this when writing the 50% in docstring:

> Percentage-wise, portion available to users:
> Memory:  `61.65%` of total memory = 20.24 GB /  32.83 GB
> CPU: `64.625%` of total vCPUs10.34 vCPU = 15.89 (allocatable) - 5.55 (daemon set in stage cluster)

Reply:
Got it, that makes sense! If possible then, I'd add a comment after `16` with `# = 32.83GB (total available) / 2 = ~16GB`.